src/test/groovy/com/google/protobuf/gradle/ProtobufJavaPluginTest.groovy

```diff
@@ -621,6 +621,96 @@ class ProtobufJavaPluginTest extends Specification {
     limit == GenerateProtoTask.DEFAULT_CMD_LENGTH_LIMIT
   }
 
+  void "test custom java executable in extension"() {
+    given: "a basic project"
+    Project project = setupBasicProject()
+
+    when: "a java executable is specified on the protobuf extension"
+    project.extensions.getByType(ProtobufExtension).javaExecutablePath.set("/custom-java.exe")
+
+    then: "all tasks get the custom executable"
+    assert project.extensions.getByType(ProtobufExtension).javaExecutablePath.get() == "/custom-java.exe"
+    assert ((GenerateProtoTask)project.tasks.generateProto).javaExecutablePath.get() == "/custom-java.exe"
+    assert ((GenerateProtoTask)project.tasks.generateTestProto).javaExecutablePath.get() == "/custom-java.exe"
+  }
+
+  void "test custom java executable in task"() {
+    given: "a basic project"
+    Project project = setupBasicProject()
+
+    when: "a java executable is specified on the generate proto task"
+    ((GenerateProtoTask)project.tasks.generateProto).javaExecutablePath.set("/custom-java.exe")
+
+    then: "generate proto task uses configured executable"
+    assert ((GenerateProtoTask)project.tasks.generateProto).javaExecutablePath.get() == "/custom-java.exe"
+
+    and: "extension and test task use default executable"
+    assert project.extensions.getByType(ProtobufExtension).javaExecutablePath
+            .get() == ProtobufExtension.computeJavaExePath()
+    assert ((GenerateProtoTask)project.tasks.generateTestProto).javaExecutablePath
+            .get() == ProtobufExtension.computeJavaExePath()
+  }
+
+  void "test custom java executable in extension and task"() {
+    given: "a basic project"
+    Project project = setupBasicProject()
+
+    when: "a java executable is specified on the protobuf extension and generate proto task"
+    project.extensions.getByType(ProtobufExtension).javaExecutablePath.set("/ext-java.exe")
+    ((GenerateProtoTask)project.tasks.generateProto).javaExecutablePath.set("/task-java.exe")
+
+    then: "extension and test task use executable specified on the extension"
+    assert project.extensions.getByType(ProtobufExtension).javaExecutablePath.get() == "/ext-java.exe"
+    assert ((GenerateProtoTask)project.tasks.generateTestProto).javaExecutablePath.get() == "/ext-java.exe"
+
+    and: "generate proto task uses executable specified on task"
+    assert ((GenerateProtoTask)project.tasks.generateProto).javaExecutablePath.get() == "/task-java.exe"
+  }
+
+  @Unroll
+  void "test proto generation fails when java executable is invalid [gradle #gradleVersion]"() {
+    given: "project from testProject"
+    File projectDir = ProtobufPluginTestHelper.projectBuilder('testProjectConfigureJavaExecutable')
```

Comment:
I don't see anything special in this project. Can you just use `testProject`?

Reply:
The main difference is having to configure the protoc-gen-grpc-kotlin plugin in order to test the use of the java executable. I tried basing this test off of testProject, but the modifications required in code were getting too to be more complex than adding a new test project.